webapp/components/compat-2021.js

```diff
@@ -461,124 +623,135 @@ class Compat2021FeatureChart extends PolymerElement {
-      //
-      // The CSV columns are:
-      //   sha, date, [product-version, product-score,]+
-
-      let csvValues = line.split(',');
-      let dataTableCells = [];
-
-      // The first datatable cell is the date. Javascript Date objects use
-      // 0-indexed months, whilst the CSV is 1-indexed, so adjust for that.
-      const dateParts = csvValues[1].split('-').map(x => parseInt(x));
-      dataTableCells.push(new Date(dateParts[0], dateParts[1] - 1, dateParts[2]));
-
-      // Now handle each of the browsers. For each there is a version column,
-      // then a score column. We use the version to create the tooltip.
-      for (let i = 2; i < csvValues.length; i += 2) {
-        const version = csvValues[i];
-        const score = parseFloat(csvValues[i + 1]);
-        const browserName = tooltipBrowserNames[(i / 2) - 1];
-        const tooltip = this.createTooltip(browserName, version, score);
-
-        dataTableCells.push(score);
-        dataTableCells.push(tooltip);
-
-        // Update the browser versions lookup table; used for the revision-diff
-        // tooltip action.
-        browserVersions[(i / 2) - 1].push(version);
-      }
-      dataTable.addRow(dataTableCells);
-    });
-
     const chart = new window.google.visualization.LineChart(div);
 
-    // Setup the tooltips to show revision diff.
+    // We define a tooltip action that can quickly show users the changelog
+    // between two subsequent versions of a browser. The goal is to help users
+    // understand why an improvement or regression may have happened - though
+    // this only exposes browser changes and not test suite changes.
+    const browserVersions = await this.dataManager.getBrowserVersions(stable);
     chart.setAction({
-      id: 'revisionDiff',
-      text: 'Show diff from previous release',
+      id: 'revisionChangelog',
+      text: 'Show browser changelog',
       action: () => {
         let selection = chart.getSelection();
         let row = selection[0].row;
         let column = selection[0].column;
 
-        // Not implemented for Firefox or Safari yet.
-        if (column !== 1) {
-          alert('Diff only supported for Chrome currently');
-          return;
-        }
-
         // Map from the selected column to the browser index. In the datatable
         // Chrome is 1, Firefox is 3, Safari is 5 => these must map to [0, 1, 2].
-        let browserIdx = (column - 1) / 2;
+        let browserIdx = Math.floor(column / 2);
 
         let version = browserVersions[browserIdx][row];
         let lastVersion = version;
         while (row > 0 && lastVersion === version) {
           row -= 1;
           lastVersion = browserVersions[browserIdx][row];
         }
-        // TODO: If row == -1, we've failed, but we should grey out the
-        // option instead in that case.
-        window.open(this.getChromeDiffUrl(lastVersion, version));
+        // TODO: If row == -1 here then we've failed.
+
+        if (browserIdx === 0) {
```

Comment:
Can we look up the browser name using `browserIdx` and switch on that?

Reply:
As with elsewhere we can and probably should, but this works for now and I don't want to spend more time refactoring this. Sorry!